lib/src/layer/polygon_layer.dart

```diff
@@ -58,133 +72,190 @@ class PolygonLayer extends StatelessWidget {
+      top: -offset.dy,
+      child: kIsWeb ? paint : RepaintBoundary(child: paint),
     );
-  }
-
-  void _fillOffsets(final List<Offset> offsets, final List<LatLng> points,
-      FlutterMapState map) {
-    final len = points.length;
-    for (var i = 0; i < len; ++i) {
-      final point = points[i];
-      final offset = map.getOffsetFromOrigin(point);
-      offsets.add(offset);
-    }
   }
 }
 
 class PolygonPainter extends CustomPainter {
-  final Polygon polygonOpt;
-  final double rotationRad;
+  final Iterable<Polygon> polygons;
+  final FlutterMapState map;
+  final double zoom;
+  final double rotation;
+
+  PolygonPainter(this.polygons, this.map)
+      : zoom = map.zoom,
+        rotation = map.rotation;
+
+  int get hash {
+    _hash ??= Object.hashAll(polygons);
+    return _hash!;
+  }
+
+  int? _hash;
 
-  PolygonPainter(this.polygonOpt, this.rotationRad);
+  List<Offset> getOffsets(List<LatLng> points) {
+    return points.map((pos) => getOffset(pos)).toList();
+  }
+
+  Offset getOffset(LatLng point) {
+    final delta = map.project(point);
+    return Offset(delta.x.toDouble(), delta.y.toDouble());
+  }
 
   @override
   void paint(Canvas canvas, Size size) {
-    if (polygonOpt.offsets.isEmpty) {
-      return;
+    var path = ui.Path();
+    var paint = Paint();
+    var borderPath = ui.Path();
+    Paint? borderPaint;
+    int? lastHash;
+
+    void drawPaths() {
+      canvas.drawPath(path, paint);
+      path = ui.Path();
+      paint = Paint();
+
+      if (borderPaint != null) {
+        canvas.drawPath(borderPath, borderPaint!);
+        borderPath = ui.Path();
+        borderPaint = null;
+      }
     }
-    final rect = Offset.zero & size;
-    _paintPolygon(canvas, rect);
-  }
 
-  void _paintBorder(Canvas canvas) {
-    if (polygonOpt.borderStrokeWidth > 0.0) {
-      final borderPaint = Paint()
-        ..color = polygonOpt.borderColor
-        ..strokeWidth = polygonOpt.borderStrokeWidth;
-
-      if (polygonOpt.isDotted) {
-        final borderRadius = (polygonOpt.borderStrokeWidth / 2);
-
-        final spacing = polygonOpt.borderStrokeWidth * 1.5;
-        _paintDottedLine(
-            canvas, polygonOpt.offsets, borderRadius, spacing, borderPaint);
-
-        if (!polygonOpt.disableHolesBorder &&
-            null != polygonOpt.holeOffsetsList) {
-          for (final offsets in polygonOpt.holeOffsetsList!) {
-            _paintDottedLine(
-                canvas, offsets, borderRadius, spacing, borderPaint);
-          }
+    for (final polygon in polygons) {
+      final offsets = getOffsets(polygon.points);
+      if (offsets.isEmpty) {
+        continue;
+      }
+
+      final hash = polygon.renderHashCode;
+      if (lastHash != null && lastHash != hash) {
+        drawPaths();
+      }
+      lastHash = hash;
+
+      final holeOffsetsList = List<List<Offset>>.generate(
+          polygon.holePointsList?.length ?? 0,
+          (i) => getOffsets(polygon.holePointsList![i]));
```

Comment:
```suggestion
          (i) => getOffsets(polygon.holePointsList![i]), growable: false);
```

Make it not growable for some potential performance gains.

Reply:
Like me some less mutability but the performance argument seems far fetched.

Reply:
When I profiled it using a growable list generator (growable: true) it only went from 8ms to about 3ms in the main getOffsets function, rather than 0.4ms. It could be a non-issue but generally non-mutable lists have significantly higher performance, or can have better performance in certain situations.

Reply:
I appreciate that you profiled that beats certainly anything that I can argue. That said, if your statement: 

>  but generally non-mutable lists have significantly higher performance

was true, that would generally strike me as surprising. Growable or not, for any given size it's just a hunk of continuous memory and on the instruction level you generally don't have a notion of immutable memory (you can make pages read-only but that's different).
Similar implementations in other languages sometimes pull off a so-called small-vector optimization but that's also generally unrelated to mutability.

I could certainly imagine the underlying implementations to be different but in that case I would consider any slowness to be a bug. Mind sharing your benchmark?